### DISCUSSION TITLE
Save 3D segmentation results of each channel directly

### DISCUSSION THREAD

**Discussion by YingdiZhang** [Project-MONAI/MONAI#5937] [Q&A] [answered]
Thanks for the great job. I am using monai in a 3D segmentation task with output channel 6. I want to save the output of each channel directly. How should I write the postprocess and save image transform?

**Comment by KumoLiu** [comment, marked as answer]
Hi @YingdiZhang, you could try to use `SplitDimd` and `SaveImaged`. Here is a simple example.
```
trans = Compose(
    [
        LoadImaged(keys='img', image_only=True, ensure_channel_first=True),
        AsDiscreted(keys='img', to_onehot=6),
        FromMetaTensord(keys='img'),
        SplitDimd(keys='img'),
        SaveImaged(keys=['img_0', 'img_1'], output_dir='.', )
])
```

Hope it can help you, thanks!

**Reply by YingdiZhang** [reply]
Thanks for your reply. I use monai 0.9.0 and write a save transform myself, but it occurs an error, and I dont know why. My code is:

self.post_trans_save = Compose(
            [
                EnsureTyped(keys="heatmap_pred"),  # 
                Invertd(
                    keys="heatmap_pred",  # invert the `heatmap_pred` data field, also support multiple fields
                    transform=self.transform,
                    orig_keys="image",
                    # get the previously applied self.transforms information on the `image` data field,
                    # then invert `pred` based on this information. we can use same info
                    # for multiple fields, also support different orig_keys for different fields
                    meta_keys="heatmap_pred_meta_dict", #"heatmap_pred_meta_dict", #    # "heatmap_pred_L1_meta_dict",
                    # key field to save inverted meta data, every item maps to `keys`
                    orig_meta_keys="image_meta_dict",  # get the meta data from `image_meta_dict` field when inverting,
                    # for example, may need the `affine` to invert `Spacingd` transform,
                    # multiple fields can use the same meta data to invert
                    meta_key_postfix="meta_dict",
                    # if `meta_keys=None`, use "{keys}_{meta_key_postfix}" as the meta key,
                    # if `orig_meta_keys=None`, use "{orig_keys}_{meta_key_postfix}",
                    # otherwise, no need this arg during inverting
                    nearest_interp=False,  # don't change the interpolation mode to "nearest" when inverting transforms
                    # to ensure a smooth output, then execute `AsDiscreted` transform
                    to_tensor=True  # convert to PyTorch Tensor after inverting
                ),
                # AsDiscreted(keys="pred", argmax=True),
                SplitChanneld(keys="heatmap_pred"),
                SaveImaged(keys="heatmap_pred_0", meta_keys="heatmap_pred_meta_dict", output_dir=self.pred_output_folder,
                           output_postfix="heatmap_L1", resample=False, output_ext='.nii.gz',
                           separate_folder=False)
            ]
        )

The error is: 
transform info of `image` is not available or no InvertibleTransform applied.
  0%|                                                                                                                                                                                | 0/18 [00:05<?, ?it/s]
Traceback (most recent call last):
  File "/home/zyd/anaconda3/envs/med/lib/python3.9/site-packages/monai/transforms/transform.py", line 89, in apply_transform
    return _apply_transform(transform, data, unpack_items)
  File "/home/zyd/anaconda3/envs/med/lib/python3.9/site-packages/monai/transforms/transform.py", line 53, in _apply_transform
    return transform(parameters)
  File "/home/zyd/anaconda3/envs/med/lib/python3.9/site-packages/monai/transforms/io/dictionary.py", line 271, in __call__
    meta_data = d[meta_key] if meta_key is not None else None
KeyError: 'heatmap_pred_meta_dict'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/data/zyd/beifen_xinsuoqu_/MedSeg/MedSeg/test/Task960Tester.py", line 663, in <module>
    t960_tester.test(
  File "/data/zyd/beifen_xinsuoqu_/MedSeg/MedSeg/test/Task960Tester.py", line 401, in test
    _ = [self.post_trans_save_heatmap_L1(i) for i in decollate_batch(val_data)]
  File "/data/zyd/beifen_xinsuoqu_/MedSeg/MedSeg/test/Task960Tester.py", line 401, in <listcomp>
    _ = [self.post_trans_save_heatmap_L1(i) for i in decollate_batch(val_data)]
  File "/home/zyd/anaconda3/envs/med/lib/python3.9/site-packages/monai/transforms/compose.py", line 173, in __call__
    input_ = apply_transform(_transform, input_, self.map_items, self.unpack_items, self.log_stats)
  File "/home/zyd/anaconda3/envs/med/lib/python3.9/site-packages/monai/transforms/transform.py", line 113, in apply_transform
    raise RuntimeError(f"applying transform {transform}") from e
RuntimeError: applying transform <monai.transforms.io.dictionary.SaveImaged object at 0x7f6b8c1f4b50>

**Reply by KumoLiu** [reply]
From the error message, seems you don't have `heatmap_pred_meta_dict`, maybe you could comment the `SaveImaged` to check what meta_dict you have in your output. Thanks!
